### PR TITLE
Jt/fixes and additions

### PR DESCRIPTION
**Pull request by jamaltoutouh**
@hembergerik It has plenty of changes. We can meet to talk about it

### CODE REVIEW

File: src/enesmble_optimization/test_ga_ensemble_generator.py [hunk 1]
```diff
@@ -10,35 +10,50 @@
 
 import pathlib
 import sys
-sys.path.insert(0, str(pathlib.Path(__file__).parent.parent.absolute())) # To change the folder path to use Lipizzaner files
+
```
[the remark this answers is not in the document]
Comment (hembergerik) [reply src/enesmble_optimization/test_ga_ensemble_generator.py:13]: Should test files be in a separate `test` folder?

File: src/lipizzaner_master.py [hunk 2]
```diff
@@ -157,85 +162,112 @@ def _terminate(self, stop_clients=True, return_code=-1):
...
-                        file.write('{}: {}\n'.format(filename, weights_generator[generator.source]))
+                    with open(os.path.join(output_dir, "mixture.yml"), "a") as file:
+                        file.write("{}: {}\n".format(filename, weights_generator[generator.source]))
 
                 for discriminator in discriminator_pop.individuals:
-                    source = discriminator.source.replace(':', '-')
-                    filename = '{}{}.pkl'.format(DISCRIMINATOR_PREFIX, source)
-                    torch.save(discriminator.genome.net.state_dict(),
-                               os.path.join(output_dir, filename))
+                    source = discriminator.source.replace(":", "-")
```
Comment (hembergerik) [inline src/lipizzaner_master.py:194]: Should this be made a method for the network to call `save`?

File: src/lipizzaner_master.py [hunk 3]
```diff
@@ -157,85 +162,112 @@ def _terminate(self, stop_clients=True, return_code=-1):
...
-                self._logger.info('Saved mixture result images of client {} to target directory {}.'
-                                  .format(node_name, output_dir))
+                dataset = MixedGeneratorDataset(
+                    generator_pop,
+                    weights_generator,
+                    self.cc.settings["master"]["score_sample_size"],
+                    self.cc.settings["trainer"]["mixture_generator_samples_mode"],
+                )
+
+                if "gaussian_" in self.cc.settings["dataloader"]["dataset_name"]:
```
Comment (hembergerik) [inline src/lipizzaner_master.py:213]: Can these be handle by default arguments to avoid controlflow here

File: src/networks/standalone_sgan/sgan.py [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+from torchvision import datasets
+from torch.autograd import Variable
+import torch.nn as nn
+import torch
+import logging
+import time
+
+from networks.standalone_sgan.score_factory import ScoreCalculatorFactory
+
+directory_or_file_name = int(time.time() * 1000)
```
Comment (hembergerik) [inline src/networks/standalone_sgan/sgan.py:16]: Use more method so it can be used as a module. E.g. specify `if __name__ == '__main__':` as entry point

File: src/networks/standalone_sgan/sgan_convolutional.py [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+from torchvision import datasets
+from torch.autograd import Variable
+import torch.nn as nn
+import torch
+import logging
+import time
+
+from networks.standalone_sgan.score_factory import ScoreCalculatorFactory
+
+directory_or_file_name = int(time.time() * 1000)
```
Comment (hembergerik) [inline src/networks/standalone_sgan/sgan_convolutional.py:16]: Use methods to simplify use as a module

File: src/data/balanced_labels_batch_sampler.py [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+    """
+    BatchSampler - from an MNIST-like dataset, samples n_classes and within
+    these classes samples (label_rate * batch_size) / n_classes number of
+    samples.
+
+    Returns batches of size batch_size
+    """
+
+    def __init__(self, dataset, num_classes, batch_size, label_rate):
+        import logging
```
[the remark this answers is not in the document]
Comment (jamaltoutouh) [reply src/data/balanced_labels_batch_sampler.py:20]: Solved!

File: src/data/balanced_labels_batch_sampler.py [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+        for label in self.labels_set:
+            np.random.shuffle(self.label_to_indices[label])
+
+        # Break down dataset into 'labeled' and 'unlabeled' datasets
+        self.labeled_data = np.array([], dtype=int)
+        self.unlabeled_data = np.array([], dtype=int)
+        total_labeled_data = int(len(self.dataset) * self.label_rate)
+        self.num_labels_per_class = int(total_labeled_data / self.num_classes)
+        for label in self.labels_set:
+            self.labeled_data = np.concatenate(
```
[the remark this answers is not in the document]
Comment (jamaltoutouh) [reply src/data/balanced_labels_batch_sampler.py:54]: Solved. Line moved

File: src/data/balanced_labels_batch_sampler.py [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+                (self.labeled_data, self.label_to_indices[label][: self.num_labels_per_class],)
+            )
+            self.unlabeled_data = np.concatenate(
+                (self.unlabeled_data, self.label_to_indices[label][self.num_labels_per_class :],)
+            )
+        np.random.shuffle(self.labeled_data)
```
[the remark this answers is not in the document]
Comment (jamaltoutouh) [reply src/data/balanced_labels_batch_sampler.py:60]: No. It is not the same. 
1st we shuffle inside the labels.
2nd we shuffle the whole data because self.labeled_data and self.unlabeled_data are sorted by labels

File: src/data/balanced_labels_batch_sampler.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (jamaltoutouh) [reply src/data/balanced_labels_batch_sampler.py:54]: That was not new code. But I have solved it

File: src/data/data_loader.py [hunk 9]
```diff
@@ -29,46 +32,61 @@ def __init__(self, dataset, use_batch=True, batch_size=100, n_batches=0, shuffle
...
+
+    def load(self, train=True):
         # Image processing
 
         # Dataset
-        dataset = self.dataset(root=os.path.join(self.cc.settings['general']['output_dir'], 'data'),
-                               train=True,
-                               transform=self.transform(),
-                               download=True)
+        if self.dataset_name == "svhn":
```
[the remark this answers is not in the document]
Comment (jamaltoutouh) [reply src/data/data_loader.py:49]: The problem is that the dataset classes provided by Pytorch have a specific parameter for SVHN dataset (i.e., a parameter named **split** which in the other datasets is named **train**). Thus, we need the if-else statements in any case. I have change it to create the dictionary name and changing the keyword name

File: src/enesmble_optimization/test_ga_ensemble_generator.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (jamaltoutouh) [reply src/enesmble_optimization/test_ga_ensemble_generator.py:13]: It is not really a test is a way to prove the code and see how it works.

File: src/lipizzaner_master.py (hunk 2 above)
Comment (jamaltoutouh) [reply src/lipizzaner_master.py:194]: We did not change that, but I created the function.

File: src/lipizzaner_master.py (hunk 3 above)
Comment (jamaltoutouh) [reply src/lipizzaner_master.py:213]: It is already handled with default arguments.
